### ISSUE TITLE
pool.members_s.get_collection() return empty,([])

### ISSUE THREAD

**Issue by susanegama** [F5Networks/f5-common-python#951]
Hi, I'm Susane and work in one of F5 customers in Brazil.
I'm using f5-sdk Python API to manipulate our pools and members in BIGIP.
So, basically I need to disable a member of pool, and the node should be available to use in another pools.
(I already make it an script that disable the node).
The problem is that the result of a call in pool.members_s.get_collection() return empty,([]).

I'm following these articles from oficial documentation, but none of them return the attribute session, the one that I want to manipulate.

http://f5-sdk.readthedocs.io/en/latest/userguide/endpoints/subcollection.html#subcollection-example
http://f5-sdk.readthedocs.io/en/latest/userguide/endpoints/subcollection_resource.html?highlight=pool.members_s.members.load

This is my code:

#!/usr/bin/python

import sys
from f5.bigip import ManagementRoot

nargs = sys.argv[4:]
arg = sys.argv[1:]

if not nargs:
    print "usage: " + sys.argv[0] + "<poolname> <nodename> <port> <enable|disable>"
    sys.exit(-1)

poolName = arg[0]
nodeName = arg[1]
port = arg[2]
action = arg[3]

if action == "enable":
   newState = "enable"
elif action == "disable":
   newState =  "user-disabled"

mgmt = ManagementRoot('BIGIPSERVER', 'USER', 'PASSWORD')
pool = mgmt.tm.ltm.pools.pool.load(partition='Common', name="'{}'".format(poolName))
members = pool.members_s.get_collection()
member = pool.members_s.members.load(partition='Common', name="'{}:{}'".format(nodeName,port))
print members
print member
print locals()

API versions:

f5-icontrol-rest (1.1.0)
f5-sdk (2.1.0)

BIP version: BIG-IP 11.6.0 Build 5.0.429 Hotfix HF5

Result of execution:

$ python change_node_pool_status.py  poolName nodeName port enable
[]
<f5.bigip.tm.ltm.pool.Members object at 0x7fc49bc46bd0>
{'member': <f5.bigip.tm.ltm.pool.Members object at 0x7fc49bc46bd0>, 'ManagementRoot': <class 'f5.bigip.ManagementRoot'>, 'nodeName': '99.99.99.99', 'poolName': 'pool_teste_monitoracao', '__builtins__': <module '__builtin__' (built-in)>, 'newState': 'enable', '__file__': 'change_node_pool_status.py', 'port': '3306', 'nargs': ['enable'], '__package__': None, 'sys': <module 'sys' (built-in)>, 'members': [], 'arg': ['pool_teste_monitoracao', '99.99.99.99', '3306', 'enable'], 'action': 'enable', '__name__': '__main__', 'mgmt': <f5.bigip.ManagementRoot object at 0x7fc49e8d7e50>, '__doc__': None, 'pool': <f5.bigip.tm.ltm.pool.Pool object at 0x7fc49bc46d50>}

I'm just abble to get status of a member, using the function Stats().

#!/usr/bin/python

import sys
import string
import json
from f5.bigip import ManagementRoot

nargs = sys.argv[3:]
arg = sys.argv[1:]

if not nargs:
    print "usage: " + sys.argv[0] + " <poolname> <nodename> <port>"
    sys.exit(-1)

poolName = arg[0]
nodeName = arg[1]
port = arg[2]

mgmt = ManagementRoot('BIGIPSERVER', 'USER', 'PASSWORD')
pool = mgmt.tm.ltm.pools.pool.load(partition='Common', name="'{}'".format(poolName))
members = pool.members_s.get_collection()
member = pool.members_s.members.load(partition='Common', name="'{}:{}'".format(nodeName,port))
member_stats = member.stats.load()
status = json.dumps(member_stats.entries['sessionStatus']['description'])
print(status.strip('"'))
~                              

Result of execution:

$python get_node_pool_status.py poolName nodeName port enable
user-disabled

Thank you in advance, and I'm sorry if my question/issue is too newbie.

**Comment by jasonrahm** [comment]
I don't see a problem with the methods on same version of sdk against 11.6:

```
>>> mgmt = ManagementRoot('172.16.15.132', 'admin', 'admin')
>>> pools = mgmt.tm.ltm.pools.get_collection()
>>> for x in pools:
...     print x.name
...     
>>> pools = mgmt.tm.ltm.pools.get_collection()
>>> for x in pools:
...     print x.name
...     
testpool
>>> pool = mgmt.tm.ltm.pools.pool.load(name='testpool')
>>> pm = pool.members_s.get_collection()
>>> for x in pm:
...     print x.name
...     
172.16.103.20:80
172.16.103.21:80
```

**Comment by wojtek0806** [comment]
@susanegama 

Hi,

I am assuming you are running on 2.x python?

Would you be able to remove those lines?
````
status = json.dumps(member_stats.entries['sessionStatus']['description'])
print(status.strip('"'))
````
and replace them with:
````
status = member_stats.entries['sessionStatus']['description']
print status
````
you can do also, but the above would be better if using python 3.x:
````
status = str(member_stats.entries['sessionStatus']['description'])
print status
````
That dictionary key you access should return unicode encoded string value, see below for example:

https://gist.github.com/wojtek0806/e7db5e95ca5f1a5d42ec0cbabf7d1021#file-example

**Comment by wojtek0806** [comment]
@susanegama

Can you paste the output and script code in gist please as it is easier to read.

Also can you advise if I understand you correctly,

The problem you described: is that you are disabling pool member and it is not present in pool members collection afterwards?

I would like to run your script locally so please post it in gist.girhub.com and paste link here

**Comment by wojtek0806** [comment]
apologies it should be gist.github.com

**Comment by susanegama** [comment]
Hi @wojtek0806 

Actually, the problem is that I could not recover the attibute "session" to alter it's state.
My code: https://gist.github.com/susanegama/334173b2e17941e500749ed56601e82d

According to documentation ( https://f5-sdk.readthedocs.io/en/latest/userguide/endpoints/subcollection_resource.html#subcollection-resource-section and https://f5-sdk.readthedocs.io/en/latest/userguide/endpoints/subcollection.html#subcollection-example), the result should be like this, returning the "session" attribute:

{
    "kind": "tm:ltm:pool:members:membersstate",
    "name": "n1:80",
    "partition": "Common",
    "fullPath": "/Common/n1:80",
    "generation": 18703,
    "selfLink": "https://localhost/mgmt/tm/ltm/pool/~Common~p1/members/~Common~n1:80?ver=11.6.0",
    "address": "192.168.51.51",
    "connectionLimit": 0,
    "dynamicRatio": 1,
    "ephemeral": "false",
    "fqdn": {
      "autopopulate": "disabled",
    }
    "inheritProfile": "enabled",
    "logging": "disabled",
    "monitor": "default",
    "priorityGroup": 0,
    "rateLimit": "disabled",
    "ratio": 1,
    **"session": "user-enabled",**
    "state": "unchecked",
}

Thank you in advance.

**Comment by susanegama** [comment]
@jasonrahm 

The attribute name I can get, but I need to get the attribute "session", and this one isn't returning, because "**members = pool.members_s.get_collection()**" is empty.

**Comment by wojtek0806** [comment]
@susanegama 

I have tested this against 11.5.4, 11.6.0 Final, 11.6.1 and 12.1.0 and I can access this attribute fine. 

See below:
https://gist.github.com/wojtek0806/e6c5280496954de01b253c5044b5c4e3

If you just do a simple GET call to /mgmt/tm/ltm/pool/~<partition>~<poolname\>/members what is the json returned? (please substitute partition and poolname to match the resource obviously)

**Comment by wojtek0806** [comment]
@susanegama 
I believe I have found the root cause of the problem, actually 2 root causes:

1. There is a slight mistake in your code:

https://gist.github.com/susanegama/334173b2e17941e500749ed56601e82d#gistcomment-1980360

2. There seems to be a bug in 11.6.0 (fixed in 11.6.1 11.5.4, 12.1.0) which allowed you to send REST calls to this URI:
````
https://ltm5:443/mgmt/tm/ltm/pool/~Common~'test_pool'
````
Your BigIP has responded with a trimmed down Json, that resulted in an empty list[]

When it works the call to the said URI(https://ltm5:443/mgmt/tm/ltm/pool/~Common~'test_pool'), should return the following error:

````

D:\VirtualEnvs\sdk_2_7\Scripts\python.exe -s C:/nodechange.py test_pool 1.1.1.1 80 enable
Traceback (most recent call last):
  File "C:/nodechange.py", line 25, in <module>
    pool = mgmt.tm.ltm.pools.pool.load(partition='Common', name="'{}'".format(poolName))
  File "C:\Users\wypior\Documents\PythonProjects\f5-common-python\f5\bigip\resource.py", line 986, in load
    return self._load(**kwargs)
  File "C:\Users\wypior\Documents\PythonProjects\f5-common-python\f5\bigip\resource.py", line 965, in _load
    response = refresh_session.get(base_uri, **kwargs)
  File "C:\Python27\Lib\site-packages\icontrol\session.py", line 272, in wrapper
    raise iControlUnexpectedHTTPError(error_message, response=response)
icontrol.exceptions.iControlUnexpectedHTTPError: 404 Unexpected Error: Not Found for uri: https://ltm6:443/mgmt/tm/ltm/pool/~Common~'test_pool'
Text: u'{"code":404,"message":"Object not found - /Common/\'test_pool\'","errorStack":[]}'

````

I have tested the corrected version, removed the extra '' you had in there and now it displays properly:

````
D:\VirtualEnvs\sdk_2_7\Scripts\python.exe -s C:/nodechange.py test_pool 1.1.1.1 80 enable

[<f5.bigip.tm.ltm.pool.Members object at 0x03405FD0>]
<f5.bigip.tm.ltm.pool.Members object at 0x03405F90>
{'member': <f5.bigip.tm.ltm.pool.Members object at 0x03405F90>, 'ManagementRoot': <class 'f5.bigip.ManagementRoot'>, 'nodeName': '1.1.1.1', 'poolName': 'test_pool', '__builtins__': <module '__builtin__' (built-in)>, 'newState': 'enable', '__file__': 'C:/nodechange.py', 'port': '80', 'nargs': ['enable'], '__package__': None, 'sys': <module 'sys' (built-in)>, 'members': [<f5.bigip.tm.ltm.pool.Members object at 0x03405FD0>], 'arg': ['test_pool', '1.1.1.1', '80', 'enable'], 'action': 'enable', '__name__': '__main__', 'mgmt': <f5.bigip.ManagementRoot object at 0x025A0870>, '__doc__': None, 'pool': <f5.bigip.tm.ltm.pool.Pool object at 0x033C69B0>}

Process finished with exit code 0
````

I hope this helps :)

**Comment by susanegama** [comment]
@wojtek0806 Thank you so much, it helped a lot.
The problem was the single quotes :)
